// n8n-editor-ui/dist/assets/useNodeViewVersionSwitcher-CXU1PYlK.js
import { K as useUIStore, $ as useCanvasStore, r as ref, q as computed, V as VIEWS, g as useI18n, at as useNDVStore, T as useWorkflowsStore, p as useSettingsStore, aw as useLocalStorage, ak as useTelemetry } from "./index-CWcTt6jd.js";
function useBeforeUnload({ route }) {
  const uiStore = useUIStore();
  const canvasStore = useCanvasStore();
  const i18n = useI18n();
  const unloadTimeout = ref(null);
  const isDemoRoute = computed(() => route.name === VIEWS.DEMO);
  const handlers = [];
  function onBeforeUnload(e) {
    if (isDemoRoute.value || window.preventNodeViewBeforeUnload) {
      return;
    }
    handlers.forEach((handler) => handler());
    if (uiStore.stateIsDirty) {
      e.returnValue = true;
      return true;
    } else {
      canvasStore.startLoading(i18n.baseText("nodeView.redirecting"));
      return;
    }
  }
  function addBeforeUnloadHandler(handler) {
    handlers.push(handler);
  }
  function addBeforeUnloadEventBindings() {
    window.addEventListener("beforeunload", onBeforeUnload);
  }
  function removeBeforeUnloadEventBindings() {
    if (unloadTimeout.value) {
      clearTimeout(unloadTimeout.value);
    }
    window.removeEventListener("beforeunload", onBeforeUnload);
  }
  return {
    onBeforeUnload,
    addBeforeUnloadEventBindings,
    removeBeforeUnloadEventBindings,
    addBeforeUnloadHandler
  };
}
function useNodeViewVersionSwitcher() {
  const ndvStore = useNDVStore();
  const workflowsStore = useWorkflowsStore();
  const telemetry = useTelemetry();
  const settingsStore = useSettingsStore();
  const isNewUser = computed(() => workflowsStore.activeWorkflows.length === 0);
  const defaultVersion = settingsStore.isCanvasV2Enabled ? "2" : "1";
  const nodeViewVersion = useLocalStorage("NodeView.version", defaultVersion);
  const nodeViewVersionMigrated = useLocalStorage("NodeView.migrated", false);
  function setNodeViewSwitcherDropdownOpened(visible) {
    if (!visible) {
      setNodeViewSwitcherDiscovered();
    }
  }
  const nodeViewSwitcherDiscovered = useLocalStorage("NodeView.switcher.discovered.beta", false);
  function setNodeViewSwitcherDiscovered() {
    nodeViewSwitcherDiscovered.value = true;
  }
  const isNodeViewDiscoveryTooltipVisible = computed(
    () => !isNewUser.value && !ndvStore.activeNodeName && nodeViewVersion.value === "2" && !nodeViewSwitcherDiscovered.value
  );
  function switchNodeViewVersion() {
    const toVersion = nodeViewVersion.value === "2" ? "1" : "2";
    if (!nodeViewVersionMigrated.value) {
      nodeViewVersionMigrated.value = true;
    }
    telemetry.track("User switched canvas version", {
      to_version: toVersion
    });
    nodeViewVersion.value = toVersion;
  }
  function migrateToNewNodeViewVersion() {
    if (nodeViewVersionMigrated.value || nodeViewVersion.value === "2") {
      return;
    }
    switchNodeViewVersion();
  }
  return {
    isNewUser,
    nodeViewVersion,
    nodeViewVersionMigrated,
    nodeViewSwitcherDiscovered,
    isNodeViewDiscoveryTooltipVisible,
    setNodeViewSwitcherDropdownOpened,
    setNodeViewSwitcherDiscovered,
    switchNodeViewVersion,
    migrateToNewNodeViewVersion
  };
}
export {
  useNodeViewVersionSwitcher as a,
  useBeforeUnload as u
};
