// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolWorkflow/v2/ToolWorkflowV2.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolWorkflowV2 = void 0;
const GenericFunctions_1 = require("n8n-nodes-base/dist/utils/workflowInputsResourceMapping/GenericFunctions");
const WorkflowToolService_1 = require("./utils/WorkflowToolService");
const versionDescription_1 = require("./versionDescription");
class ToolWorkflowV2 {
    constructor(baseDescription) {
        this.methods = {
            localResourceMapping: {
                loadWorkflowInputMappings: GenericFunctions_1.loadWorkflowInputMappings,
            },
        };
        this.description = {
            ...baseDescription,
            ...versionDescription_1.versionDescription,
        };
    }
    async supplyData(itemIndex) {
        const workflowToolService = new WorkflowToolService_1.WorkflowToolService(this);
        const name = this.getNodeParameter('name', itemIndex);
        const description = this.getNodeParameter('description', itemIndex);
        const tool = await workflowToolService.createTool({
            name,
            description,
            itemIndex,
        });
        return { response: tool };
    }
}
exports.ToolWorkflowV2 = ToolWorkflowV2;
